pkg/auth/dependency/interaction/interaction.go

```diff
@@ -1,34 +1,71 @@
 package interaction
 
 import (
+	"encoding/json"
 	"time"
 
-	"github.com/skygeario/skygear-server/pkg/core/authn"
 	"github.com/skygeario/skygear-server/pkg/core/skyerr"
 )
 
+type TypeID struct {
+	Type string `json:"type"`
+	ID   string `json:"id"`
+}
+
 // Interaction represents an interaction with authenticators/identities, and authentication process.
 type Interaction struct {
-	Token       string            `json:"token"`
-	CreatedAt   time.Time         `json:"created_at"`
-	ExpireAt    time.Time         `json:"expire_at"`
-	SessionID   string            `json:"session_id,omitempty"`
-	SessionType authn.SessionType `json:"session_type,omitempty"`
-	ClientID    string            `json:"client_id,omitempty"`
-
-	Intent Intent           `json:"intent"`
+	Token     string    `json:"token"`
+	CreatedAt time.Time `json:"created_at"`
+	ExpireAt  time.Time `json:"expire_at"`
+	Session   *TypeID   `json:"session,omitempty"`
+	ClientID  string    `json:"client_id,omitempty"`
+
+	Intent Intent           `json:"-"`
 	Error  *skyerr.APIError `json:"error,omitempty"`
 
 	UserID                   string             `json:"user_id"`
-	IdentityID               string             `json:"identity_id,omitempty"`
+	IdentityID               *TypeID            `json:"identity,omitempty"`
 	Identity                 *IdentityInfo      `json:"-"`
-	PrimaryAuthenticatorID   string             `json:"primary_authenticator_id,omitempty"`
+	PrimaryAuthenticatorID   *TypeID            `json:"primary_authenticator,omitempty"`
 	PrimaryAuthenticator     *AuthenticatorInfo `json:"-"`
-	SecondaryAuthenticatorID string             `json:"secondary_authenticator_id,omitempty"`
+	SecondaryAuthenticatorID *TypeID            `json:"secondary_authenticator,omitempty"`
 	SecondaryAuthenticator   *AuthenticatorInfo `json:"-"`
 
 	PendingIdentity      *IdentityInfo       `json:"pending_identity,omitempty"`
 	PendingAuthenticator *AuthenticatorInfo  `json:"pending_authenticator,omitempty"`
 	NewIdentiies         []IdentityInfo      `json:"new_identities,omitempty"`
 	NewAuthenticators    []AuthenticatorInfo `json:"new_authenticators,omitempty"`
 }
+
+func (i *Interaction) MarshalJSON() ([]byte, error) {
```

Comment:
We still need to use custom marshal methods, since intent in `Interaction` struct is an interface and we need to instantiate the correct struct at deserialization.